File: TPCPhotoBrowserView/TPCPhotoBrowserViewObjective-C/TPCImageView.m
//
//  TPCImageView.m
//  TPCPhotoBrowserView
//
//  Created by tripleCC on 15/12/9.
//  Copyright © 2015年 tripleCC. All rights reserved.
//

#import "TPCImageView.h"
#import "DALabeledCircularProgressView.h"
#import "UIImageView+WebCache.h"

@interface TPCImageView() <UIScrollViewDelegate>
{
    UIScrollView *_scrollView;
    DALabeledCircularProgressView *_progressView;
}
@end

@implementation TPCImageView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setupSubviews];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        [self setupSubviews];
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    [self adjustImageFrame];
}

- (void)adjustImageFrame {
    CGSize boundsSize = self.bounds.size;
    CGRect frameToCenter = _imageView.frame;
    
    if (frameToCenter.size.width < boundsSize.width) {
        frameToCenter.origin.x = (boundsSize.width - frameToCenter.size.width) * 0.5;
    } else {
        frameToCenter.origin.x = 0;
    }
    
    if (frameToCenter.size.height < boundsSize.height) {
        frameToCenter.origin.y = (boundsSize.height - frameToCenter.size.height) * 0.5;
    } else {
        frameToCenter.origin.y = 0;
    }
    
    if (!CGRectEqualToRect(_imageView.frame, frameToCenter)) {
        _imageView.frame = frameToCenter;
    }
}

- (void)setImageURLString:(NSString *)imageURLString {
    if ([_imageURLString isEqualToString:imageURLString]) { return; }
    _imageURLString = imageURLString;
    [self resetSubviews];
    [_imageView sd_setImageWithURL:[NSURL URLWithString:imageURLString] placeholderImage:nil options:SDWebImageRetryFailed progress:^(NSInteger receivedSize, NSInteger expectedSize) {
        CGFloat progress = (CGFloat)receivedSize / (CGFloat)expectedSize;
        if (progress < 0.009) {
            _progressView.progressLabel.text = @"0.00";
            _progressView.progress = 0;
        } else {
            [_progressView setProgress:progress animated:YES];
            _progressView.progressLabel.text = [NSString stringWithFormat:@"%.2f", _progressView.progress];
            if (progress >= 1.0) {
                [UIView animateWithDuration:0.1 animations:^{
                    _progressView.alpha = 0;
                }];
            }
        }
    } completed:^(UIImage *image, NSError *error, SDImageCacheType cacheType, NSURL *imageURL) {
        [self adjustImageViewFrameByImage:image];
        _progressView.alpha = 0;
    }];
}

- (void)setupSubviews {
    _scrollView = [[UIScrollView alloc] initWithFrame:self.bounds];
    _scrollView.delegate = self;
    _scrollView.maximumZoomScale = 3.0;
    _scrollView.minimumZoomScale = 1.0;
    _scrollView.showsHorizontalScrollIndicator = NO;
    _scrollView.showsVerticalScrollIndicator = NO;
    _scrollView.backgroundColor = [UIColor blackColor];
    [self addSubview:_scrollView];
    
    _imageView = [[UIImageView alloc] init];
    _imageView.userInteractionEnabled = YES;
    _imageView.clipsToBounds = YES;
    _imageView.contentMode = UIViewContentModeScaleAspectFit;
    [_scrollView addSubview:_imageView];
    
    UITapGestureRecognizer *doubleTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(doubleTap:)];
    doubleTap.numberOfTapsRequired = 2;
    [_imageView addGestureRecognizer:doubleTap];
    
    _progressView = [[DALabeledCircularProgressView alloc] initWithFrame:CGRectMake(0, 0, 40, 40)];
    _progressView.center = CGPointMake(self.bounds.size.width * 0.5, self.bounds.size.height * 0.5);
    _progressView.roundedCorners = YES;
    _progressView.progressLabel.font = [UIFont systemFontOfSize:10.0];
    _progressView.progressLabel.textColor = [UIColor whiteColor];
    [self addSubview:_progressView];
}

- (void)doubleTap: (UIGestureRecognizer *)gesture {
    if (_scrollView.zoomScale == 1) {
        CGPoint point = [gesture locationInView:gesture.view];
        CGFloat width = self.bounds.size.width / _scrollView.maximumZoomScale;
        CGFloat height = self.bounds.size.height / _scrollView.maximumZoomScale;
        [_scrollView zoomToRect:CGRectMake(point.x - width * 0.5, point.y - height * 0.5, width, height) animated:YES];
    } else {
        [_scrollView setZoomScale:1 animated:YES];
    }
}

- (void)adjustImageViewFrameByImage:(UIImage *)image {
    _imageView.bounds = CGRectMake(0, 0, image.size.width, image.size.height);
    CGFloat xScale = self.bounds.size.width / image.size.width;
    CGFloat yScale = self.bounds.size.height / image.size.height;
    CGFloat minScale = MIN(xScale, yScale);
    _imageView.bounds = CGRectMake(0, 0, _imageView.bounds.size.width * minScale, _imageView.bounds.size.height * minScale);
    [self adjustImageFrame];
}

- (void)resetSubviews {
    _scrollView.contentInset = UIEdgeInsetsZero;
    _scrollView.contentOffset = CGPointZero;
    _scrollView.contentSize = CGSizeZero;
    _imageView.transform = CGAffineTransformIdentity;
    _progressView.alpha = 1;
    _progressView.progress = 0;
    _progressView.progressLabel.text = @"0.00";
}

#pragma mark UIScrollViewDelegate
- (UIView *)viewForZoomingInScrollView:(UIScrollView *)scrollView {
    return _imageView;
}

- (void)scrollViewDidZoom:(UIScrollView *)scrollView {
    [self adjustImageFrame];
}
@end

File: TPCPhotoBrowserView/TPCPhotoBrowserViewSwift/TPCImageView.swift
//
//  TPCImageView.swift
//  TPCPageScrollView
//
//  Created by tripleCC on 15/11/25.
//  Copyright © 2015年 tripleCC. All rights reserved.
//

import UIKit
import DACircularProgress
import SDWebImage

class TPCImageView: UIView {
    var imageMode: UIViewContentMode = UIViewContentMode.ScaleAspectFit {
        didSet {
            imageView.contentMode = imageMode
        }
    }
    private var scrollView: UIScrollView!
    private var imageView: UIImageView!
    private var progressView: DALabeledCircularProgressView!
    var imageURLString: String! {
        didSet {
            resetSubviews()
            imageView.sd_setImageWithURL(NSURL(string: imageURLString), placeholderImage: UIImage(), options: SDWebImageOptions.RetryFailed, progress: { (receivedSize, expectedSize) -> Void in
                guard CGFloat(receivedSize) / CGFloat(expectedSize) > 0.009 else {
                    self.progressView.progressLabel.text = "0.00"
                    self.progressView.progress = 0
                    return
                }
                
                self.progressView.setProgress(CGFloat(receivedSize) / CGFloat(expectedSize), animated: true)
                self.progressView.progressLabel.text = String(format:"%.2f", self.progressView.progress)
                if receivedSize / expectedSize >= 1 {
                    UIView.animateWithDuration(0.1, animations: { () -> Void in
                        self.progressView.alpha = 0
                    })
                }
                }) { (image, error, cacheType, imageURL) -> Void in
                    self.adjustImageViewFrameByImage(image)
                    self.progressView.alpha = 0
            }
        }
    }
    private func adjustImageViewFrameByImage(image: UIImage?) {
        if let image = image {
            if imageView.contentMode == UIViewContentMode.ScaleAspectFit || imageView.contentMode == UIViewContentMode.Center{
                if (image.size.width > image.size.height) {
                    imageView.frame = CGRect(x: 0, y: 0, width: bounds.width, height: bounds.width / image.size.width * image.size.height)
                } else {
                    imageView.frame = CGRect(x: 0, y: 0, width: bounds.height / image.size.height * image.size.width, height: image.size.height)
                }
            } else {
                imageView.frame = bounds
            }
            setNeedsLayout()
            layoutIfNeeded()
        }
    }
    
    private func resetSubviews() {
        scrollView.contentInset = UIEdgeInsetsZero
        scrollView.contentOffset = CGPointZero
        scrollView.contentSize = CGSizeZero
        imageView.transform = CGAffineTransformIdentity
        progressView.alpha = 1
        progressView.progress = 0
        progressView.progressLabel.text = "0.00"
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupSubviews()
    }
    
    private func setupSubviews() {
        scrollView = UIScrollView()
        scrollView.delegate = self
        scrollView.frame = bounds
        scrollView.maximumZoomScale = 3.0
        scrollView.minimumZoomScale = 1.0
        scrollView.showsHorizontalScrollIndicator = false
        scrollView.showsVerticalScrollIndicator = false
        scrollView.backgroundColor = UIColor.blackColor()
        addSubview(scrollView)
        
        imageView = UIImageView()
        imageView.userInteractionEnabled = true
        imageView.clipsToBounds = true
        imageView.contentMode = UIViewContentMode.ScaleAspectFit
        scrollView.addSubview(imageView)
        
        let doubleTap = UITapGestureRecognizer(target: self, action: "doubleTap:")
        doubleTap.numberOfTapsRequired = 2
        imageView.addGestureRecognizer(doubleTap)
        
        progressView = DALabeledCircularProgressView(frame: CGRect(x: 0, y: 0, width: 40, height: 40))
        progressView.center = CGPoint(x: bounds.width * 0.5, y: bounds.height * 0.5)
        progressView.roundedCorners = Int(true)
        progressView.progressLabel.font = UIFont.systemFontOfSize(10)
        progressView.progressLabel.textColor = UIColor.whiteColor()
        addSubview(progressView)
    }

    func doubleTap(gesture: UIGestureRecognizer) {
        if scrollView.zoomScale == 1 {
            let point = gesture.locationInView(gesture.view)
            let width = bounds.width / scrollView.maximumZoomScale
            let height = bounds.height / scrollView.maximumZoomScale
            scrollView.zoomToRect(CGRect(x: point.x - width * 0.5, y: point.y - height * 0.5, width: width, height: height), animated: true)
        } else {
            scrollView.setZoomScale(scrollView.minimumZoomScale, animated: true)
        }
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        
        let boundsSize = bounds.size
        var frameToCenter = imageView.frame
        
        if frameToCenter.size.width < boundsSize.width {
            frameToCenter.origin.x = (boundsSize.width - frameToCenter.width) * 0.5
        } else {
            frameToCenter.origin.x = 0
        }
        
        if frameToCenter.size.height < boundsSize.height {
            frameToCenter.origin.y = (boundsSize.height - frameToCenter.height) * 0.5
        } else {
            frameToCenter.origin.y = 0
        }
        if !CGRectEqualToRect(imageView.frame, frameToCenter) {
            imageView.frame = frameToCenter
        }
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

extension TPCImageView: UIScrollViewDelegate {
    func viewForZoomingInScrollView(scrollView: UIScrollView) -> UIView? {
        return imageView
    }
    func scrollViewDidZoom(scrollView: UIScrollView) {
        setNeedsLayout()
        layoutIfNeeded()
    }
}